05-rx-js/src/services/utils.js

```diff
@@ -0,0 +1,31 @@
+import { HttpService as httpService } from "./http.service";
+import { map } from "rxjs/operators";
+
+export function loadPosts() {
+    return httpService.getData('https://jsonplaceholder.typicode.com/posts').pipe(
+        map((data) => convertPostsToMarkup(data))
+    );
+}
+
+export function filterBySelector(query, parentElem) {
+    parentElem.innerHTML = '';
+    httpService.getData('https://jsonplaceholder.typicode.com/posts').pipe(
+        map((data) => data.filter(
+            post => post.title.includes(query)
+        )),
+        map((data) => convertPostsToMarkup(data))
+    ).subscribe((markUp) => parentElem.innerHTML = markUp);
+}
+
+
+function convertPostsToMarkup(posts) {
+    let postsMarkup = '';
+    posts.forEach(post => {
+        postsMarkup += `
+                    <div>
+                        <h2>${post.title}</h2>
+                        <p>${post.body}</p>
+                    </div>`;
+    });
+    return postsMarkup;
+}
```

Comment:
Since you're good in ES6 you could have this utils written like so:
```js
export const convertPostsToMarkup = posts =>
    posts.reduce((acc, cur) =>
        `${acc}
            <div>
                <h2>${cur.title}</h2>
                <p>${cur.body}</p>
            </div>`
        , '')
```

Try the following:
- Only have `convertPostsToMarkup` exported from utils
- Have httpService exported as plain object, with the specific URL used inside the `getData` method (since it never changes for the method)
- Have httpService imported to `main`
- Use to separate `Subject`s in `main`: one for the full list, and another one for the filteredList.
- Load button click would getData() and set next() for both: full list and filteredList.
- Filter input change would have to use the `fullList` Subject together with the input to produce the next value of the `filteredList` Subject (how can you can the latest value from another Subject?).
- You would only use the `convertPostsToMarkup` once, when `filteredList` changes

This way you would have all your app built based on observables, and remove all the duplications, especially the heavy re-loads of the data.